text/0003-webcontents-invoke.md

```diff
@@ -0,0 +1,377 @@
+
+If `timeout` is not provided, a default value of 30000 milliseconds is used. This means
+that if the function doesn't receive a response within 30 seconds, it will time out and throw an error.
+
+The renderer process should listen for channel with `ipcRenderer.handle()`.
+
+For example:
+
+```
+// main process
+const { BrowserWindow } = require('electron');
+
+const rendererWindow = new BrowserWindow({ ... });
+
+rendererWindow.webContents.on('did-finish-load', async () => {
+  const result = await rendererWindow.webContents.invoke('my-channel', arg1, arg2);
+  console.log(result); // Output from the renderer process
+
+  // adding custom timeout
+  const result = await rendererWindow.webContents.invoke({timeout: 1000}, 'my-channel', arg1, arg2);
+});
+```
+
+### `ipcRenderer.handle(channel, ...args)`
+- `channel` string
+- `listener` Function<Promise<any\> | any\>
+- `event` IpcRendererInvokeEvent
+- `...args` any[]
+
+Handles a single `invoke`able IPC message, then removes the listener. See `ipcRenderer.handle(channel, listener)`.
+
+For example:
+
+```
+// renderer process
+const { ipcRenderer } = require('electron');
+
+ipcRenderer.handle('my-channel', async (event, ...args) => {
+  const [arg1, arg2] = args;
+  console.log(`Received from main process: ${arg1} ${arg2}`);
+  return 'Response from renderer process';
+});
+```
+
+This feature will impact existing use cases by providing a more streamlined approach to communication between the main and renderer processes. Developers can migrate directly to utilize the new APIs. Additionally, the existing IPC messaging methods will remain available for backward compatibility.
+The proposed feature will make Electron code more maintainable by providing a consistent API for communication between the main and renderer processes. The promise-based pattern and the ability to handle events directly in the main process will reduce boilerplate code and improve code readability.
+
+## Reference-level explanation
+
+The `webContents.invoke()` will simply be a wrapper around `mainFrame.invoke`. The javascript wrapper for `WebFrameMain.invoke` will handle the timeout and reject the promise if timeout is exceeded. Default = 30000 ms.
+```
+// lib/browser/api/web-frame-main.ts
+
+WebFrameMain.prototype.invoke = async function (...args) {
+  // Check if the first argument is an options object
+  const isOptionsProvided = typeof args[0] === 'object';
+  const { timeout = DEFAULT_TIMEOUT } = isOptionsProvided ? args.shift() : {};
+  const channel = args.shift();
+
+  if (typeof channel !== 'string') {
+    throw new TypeError('Missing required channel argument');
+  }
+
+  if (typeof timeout !== 'number') {
+    throw new TypeError('Invalid timeout argument');
+  }
+
+  return new Promise((resolve, reject) => {
+    const timeoutId = setTimeout(() => {
+      reject(new Error(`Timeout after ${timeout}ms waiting for IPC response on channel '${channel}'`));
```

Comment:
NB, this will reject the promise but won't tear down any of the Mojo state. There will still be a receiver entry on the C++ side until a response is received or the mojo pipe is disconnected. This seems like a memory leak to me, and I'm not sure there's a straightforward way to resolve it as I don't think the C++ mojo bindings permit "cancelling" an IPC request.

This wouldn't necessarily be a problem if timeouts always happened in the case of a "broken" renderer (e.g. crashed, hung, mid-navigation, etc.) as we would expect the pipe to be torn down in short order anyway. However, a timeout can happen if the renderer side code drops the promise or otherwise fails to respond on the user-side. I feel like that shouldn't result in leaked memory in the main process.

Reply:
Thanks for reviewing and calling out the potential memory leak! This is my first time working with cross-process APIs and C++ so your feedback is helpful. 

Brainstorming some ways we can manage mojo state in case of crashes or dropped promises [based on these best practices](https://chromium.googlesource.com/chromium/src/+/HEAD/mojo/public/cpp/bindings/README.md#best-practices-for-dealing-with-process-crashes-and-callbacks): 

- Renderer: Mojo pipe will be disconnected and cleaned up via receiver handler `receiver_.set_disconnect_handler`
- Main: Can we add [mojo::WrapCallbackWithDropHandler](https://cs.chromium.org/chromium/src/mojo/public/cpp/bindings/callback_helpers.h?l=46) to clean up state?

What exactly will be the state that will cause the memory leak? I'm assuming that the `WebFrameMain` wants to maintain an active connection to the associated renderer process, even if a specific IPC request times out or fail to respond. So does calling `TeardownMojoConnection` make sense if we still want to keep the connection?

If we are not trying to call `TeardownMojoConnection` and there is still potential memory leak caused by the individual IPC request on the main process, can we include logic in `mojo::WrapCallbackWithDropHandler` to tear it down?

Thanks for the guidance!